### ISSUE TITLE
U+2060 (Word Joiner) warns "fontselect: failed to find any fallback for font: (sans-serif, 400, 0)"

### ISSUE THREAD

**Issue by avih** [libass/libass#507]
I don't know if this is a libass issue or mpv issue, but I'd appreciate your help to recognize the source of the issue and/or the best way to address it.

Platform: Windows 10
mpv: master
libass: d82f999 (2021-05-01)

Whenever mpv tries to display backslash at the OSD, the backslash IS displayed, but mpv prints a warning.

Steps to reprodue:
- run `mpv --no-config --force-window --idle`
- press tilde/backtick at the window to open the internal console
- type `show-text foo\bar` and press enter (the backslash is correctly taken literally - that's how mpv parsing works)

Result: `foo\bar` is displayed at the OSD, and the terminal displays
```
[osd/libass] fontselect: failed to find any fallback for font: (sans-serif, 400, 0)
```

I tried other fonts by adding to the mpv command line e.g. `--osd-font=Arial` or `--osd-font="Comic Sans MS"` - with same warning except the different font name at the message (the display font does change visibly).

mpv does escape `\` internally by default for any OSD text by appending U+2060 right after it so that libass doesn't recognize it as a begining of an ass sequence, via this code: https://github.com/mpv-player/mpv/blob/dd4d239bcb151b08eb81b38486515a3e1222e1f5/sub/osd_libass.c#L223-L225

If I change `0x2060` (Word Joiner) to `0x200B` (Zero Width Space) then it _seems_ to behave the same but without the warning - i.e. this seem to fix the issue.

So I have few questions:
1. It seems to me that libass tries to find the glyph for U+2060 and fails. Is this a correct behavior? Are fonts expected to have a glyph for this codepoint?
2. For escaping `\` in strings, which should be more correct to use between U+2060 and U+200B ? (i.e. should we change mpv to use U+200B?)
3. If we do change mpv to use U+200B, could any issue arrise with older versions of libass which would be avoided with U+2060?

Thanks in advance.

**Comment by astiob** [comment]
>     1. It seems to me that libass tries to find the glyph for U+2060 and fails. Is this a correct behavior?

Per https://github.com/libass/libass/issues/71#issuecomment-61705139, this is _not_ correct behaviour. We’d like to fix it, but… uh… nobody has gotten around to doing it and we probably keep forgetting about it. Having a separate ticket tracking just this issue should help improve visibility, so thanks for reporting this!

**Comment by avih** [comment]
Thanks.

Should we change the mpv escaping to use U+200B anyway? is one of them more correct than the other in the context of libass (sans bugs)? Would U+200B have issues with older versions of libass which U+2060 wouldn't have?

**Comment by avih** [comment]
This is weird. `2060` appears only once at the libass code - in exactly the same place where `200b` is (which also only has one instance at the libass code), and both are excluded equally as far as I can tell (lines 879 and 880): https://github.com/libass/libass/blob/5447214643eacef71776350e779adf4b6c07bb3b/libass/ass_shaper.c#L876-L891

How come one produces a warning but the other doesn't? Maybe the exclusion doesn't work in some cases, but somehow fonts do have a glyph for U+200B but not for U+2060?

**Comment by astiob** [comment]
Right now, I believe there should be no difference between the rendering of U+2060 and U+200B in libass, other than the potentially different amount of font warnings in the log (which, in general, can go either way depending on the font used).

At some point in the future, we’ll add optional Unicode-aware line wrapping (#486), which I imagine mpv will likely want to enable for its OSD. Then, U+2060 will be the better one, because (if I understand correctly) U+200B would allow a line break after the backslash, which is normally forbidden if the backslash is followed by letters, ideographs, numbers or brackets.

> I _assume_ that in the past this warning was not displayed for U+2060, at least on the system of whoever added it. I.e. that this escaping was tested to behave correctly.

Since e6c71d8a94172a11a17518387ae5f3db540490c8 in 2012, the presence of the warning depends on two things:
* whether the font has a glyph for U+2060 (or U+200B),
* whether libass is using HarfBuzz.

If libass isn’t using HarfBuzz, `ass_shaper_skip_characters` (which you quoted above) is run and removes both U+2060 and U+200B from the text before attempting to find them in the font, so there’s no warning.

If libass is using HarfBuzz but the font contains a glyph, there’s no warning.

If libass is using HarfBuzz and the font lacks a glyph, the warning is logged.

**Comment by astiob** [comment]
By the way: the backslash only needs escaping if it is immediately followed by U+007B LEFT CURLY BRACKET `{`. In fact (since 3199a876dd5da0463fd6c6c231db7c3bb223c2ef in 2014), only if it is immediately followed by `{` and then at some later point by a `}`.

**Comment by astiob** [comment]
Oh… or by a `N` or `n` or `h`, because `\N` `\n` `\h` are symbol escapes for newline and NBSP. Right.

**Comment by astiob** [comment]
> So, there's no harm in simply escaping all backslashes (i.e. what mpv does now)

Oh, it’s certainly safe, although when we introduce Unicode line breaking, it might accidentally forbid an occasional legal line break. This could just be a somewhat cheap way of cutting down on the font warnings.

> instead of trying to keep up to date with what libass can interpret internally, right?

For what it’s worth, we definitely don’t plan on expanding this list, not without some kind of feature toggle. This would be a breaking change for a lot of things, mpv OSD included.

> So is this an issue of Harfbuzz not "tagging" it correctly so that libass won't try to find a glyph? Or is it an issue of libass which includes codepoints which it know should not be attempted to resolve to glyphs? Or something else?

It’s libass’s fault, although a truly optimal solution might involve some cooperation with HarfBuzz. At the moment, libass resolves the glyphs before ever calling HarfBuzz, so it’s the second thing you listed: we “should know” that these codepoints don’t need glyphs, but we’re still trying to find glyphs for them.

**Comment by astiob** [comment]
> Also, I believe HarfBuzz is mandatory currently, so I presume this condition only applies to older libass builds earlier than this mandatory requirement, yes?

Practically speaking, yes.

But technically, you can still disable HarfBuzz at runtime by calling `ass_set_shaper(renderer, ASS_SHAPING_SIMPLE)`. For subtitles, mpv exposes this as a command-line option if I recall correctly.

**Comment by avih** [comment]
> Oh, it’s certainly safe, although when we introduce Unicode line breaking, it might accidentally forbid an occasional legal line break. This could just be a somewhat cheap way of cutting down on the font warnings.

I can't say that I 100% get to the bottom of this case, but I take your word :) We might need to revisit mpv escaping once Unicode line breaking is introduced (and then we _will_ have to get to the bottom of it...).

> At some point in the future, we’ll add optional Unicode-aware line wrapping (#486), which I imagine mpv will likely want to enable for its OSD. Then, U+2060 will be the better one, because (if I understand correctly) U+200B would allow a line break after the backslash, which is normally forbidden if the backslash is followed by letters, ideographs, numbers or brackets.

Interesting. Agreed. I didn't take line-breaking rules into account. Much appreciated.

> But technically, you can still disable HarfBuzz at runtime by calling `ass_set_shaper(renderer, ASS_SHAPING_SIMPLE)`. For subtitles, mpv exposes this as a command-line option if I recall correctly.

Right. Thanks for remingind me :)

>  so it’s the second thing you listed: we “should know” that these codepoints don’t need glyphs, but we’re still trying to find glyphs for them

Right. Mind pointing me to that code? Maybe we should reuse the exclusion code I quoted earlier also for the HarfBuzz code in some way? I don't mind trying to have a go (I did try, earlier, and stopped once I realized these codepoints seem to be excluded equally).

**Comment by astiob** [comment]
> At the moment, libass resolves the glyphs before ever calling HarfBuzz

Hmm, this is not actually true. So perhaps we can in some way detect which glyphs HarfBuzz skipped over. I’ll need to dig deeper into this code to remember how exactly it works.

**Comment by astiob** [comment]
Aha, there it is. That line was actually kinda right… but now it seems as though you should be getting the warning even without HarfBuzz. Hmm.

> Mind pointing me to that code?

The overall pipeline is `ass_render_event` in `ass_render.c`. The relevant calls are:
1. `parse_events` reads the text,
2. `ass_shaper_find_runs` **selects the font** for each code point and splits them into same-font-property runs,
3. `ass_shaper_shape` runs HarfBuzz, which requests font glyphs via callbacks and receives our previously found glyphs. I assume it doesn’t request a glyph for U+200B or U+2060.

**Comment by avih** [comment]
> The overall pipeline is

Thanks. I'll try to have a loook at this code.

So just to summurize so far (please confirm):
- When libass doesn't use Harfbuzz (older libass built without HarfBuzz, or mpv with option to choose simple shaping) then both U+2060 and U+200B are excluded equally, and the warning is avoided.
- The warning does show when HarfBuzz is used (the default in mpv if libass is built with HarfBuzz support - which is mandatory since recently), because libass tries to resolve glyphs for codepoints which it knows should not be resolved.
- It warns for U+2060 but not for U+200B (I experimented with this change in mpv) because all the fonts which were attempted (`Arial`, `Comic Sans MS`, and the default `sans-serif`) do have a glyph for U+200B but don't have a glyph for U+2060?
- When escaping backslash in strings, U+2060 is better than U+200B due to Unicode line breaking rules (which won't apply until libass introduces Unicode line breaking) - which (typically?) don't allow line break when backslash is followed by most(?) codepoints.
- Once Unicode line breaking is introduced in libass, mpv might need to revisit the unconditional escaping of backslash by adding U+2060 because this might interfere incorrectly with some legal line breaks.

And maybe bypassing all these issues and questions: is there a better way to escape backslash in libass other than artificially inserting some non-printable codepoint after backslash?

**Comment by astiob** [comment]
> When libass doesn't use Harfbuzz (older libass built without HarfBuzz, or mpv with option to choose simple shaping) then both U+2060 and U+200B are excluded equally, and the warning is avoided.

Nope! Turns out the warning is logged even without HarfBuzz. I’ve just verified this in mpv with `--sub-ass-shaper=simple`.

So I don’t know why you weren’t seeing this warning before, unless you’ve changed your font.

> It warns for U+2060 but not for U+200B (I experimented with this change in mpv) because all the fonts which were attempted (`Arial`, `Comic Sans MS`, and the default `sans-serif`) do have a glyph for U+200B but don't have a glyph for U+2060?

Yes.

For what it’s worth, on my system, Comic Sans MS has neither of those glyphs, Arial has U+200B but not U+2060, and Helvetica has both.

> Once Unicode line breaking is introduced in libass, mpv might need to revisit the unconditional escaping of backslash by adding U+2060 because this might interfere incorrectly with some legal line breaks.

Yes, but that’s still safe—just suboptimal.

> is there a better way to escape backslash in libass other than artificially inserting some non-printable codepoint after backslash?

Nope. We have an issue somewhere on the tracker for this, but there’s no easy or backwards-compatible way to add this.

**Comment by avih** [comment]
> Nope! Turns out the warning is logged even without HarfBuzz. I’ve just verified this in mpv with --sub-ass-shaper=simple.
> So I don’t know why you weren’t seeing this warning before

Hmm.. maybe I'm not remembering correctly then. Either way, not too important I guess, because we can only change the future - not the past.

> For what it’s worth, on my system, Comic Sans MS has neither of those glyphs, Arial has U+200B but not U+2060, and Helvetica has both.

Interesting. I'll admit that testing with the non-default font were less thorough, so probably my fault in this summary. I assume on my system this is the same or similar. Either way, the case that some fonts have U+200B but not U+2060 can happen, and explain the difference. Thanks.

> The warning you’re quoting is “failed to find any fallback”, when none of your fonts contain this glyph (or at least libass failed to find any). This is logged at the MSGL_WARN level.

Right, I sort of get it (but I assume my Helvetica would have it same as yours, so.. whatever). Anyway, right, somewhat besides the main issue here.

> > is there a better way to escape backslash in libass other than artificially inserting some non-printable codepoint after backslash?
> 
> Nope. We have an issue somewhere on the tracker for this, but there’s no easy or backwards-compatible way to add this.

For reference, mpv internally uses a "trick" to communicate ass escaping between different parts at the code (e.g. when an mpv script command wants to display some text but only wants some of it to be ass-escaped): for an abitrary input string, all characters between the byte `0xfe` and `0xfd` (both invalid bytes in UTF-8) go through mpv ass escaping.

This ass escaping adds backslash before (source) `{`, and adds U+2060 after (source) backslash.

The strings which mpv uses with libass don't include these invalid bytes.

Maybe libass could publish some official way or procedure to escape arbitrary (parts of) strings?

Maybe some API for this? or a trick similar to what mpv uses? (maybe with some valid but reserved codepoints which can be used "legally" for such things?)

**Comment by avih** [comment]
> for an abitrary input string, all characters between the byte `0xfe` and `0xfd` (both invalid bytes in UTF-8) go through mpv ass escaping.

The above is still correct, but works the other way around: it's escaping by default (as described), where `0xfd` turns escaping off, and `0xfe` turns it back on.

> but this happens after looking them up in the font, so there’s still a warning.

Right, indeed I tested my patch also with `--sub-ass-shaper=simple` and it results in a warning. I'll address that. Thanks for clarifying.

**Comment by astiob** [comment]
While reviewing #508, I’ve discovered a completely different code point that I now believe is the optimal one to use to escape backslash and brace:

U+034F COMBINING GRAPHEME JOINER

It’s invisible. It doesn’t affect spacing, line breaking or ligatures. It almost has no impact at all, and specifically around `\` and `{` I don’t think it has _any_ impact.

**Comment by astiob** [comment]
And a correction! 😅 Despite https://unicode.org/faq/char_combmark.html and other material saying it “has no impact on line breaking”, Unicode’s Line Breaking Properties for the [Unicode Line Breaking Algorithm](https://unicode.org/reports/tr14/) classify it as glue _preventing line breaks on either side._ Oh well.

**Comment by avih** [comment]
> classify it as glue _preventing line breaks on either side._

:)

**Comment by avih** [comment]
You know, now that I think about it, maybe the best way to avoid the warning is to simply suppress it (or make info) if we couldn't find a font for a control codepoint?

On one hand this will 100% guarantee no regressions.

Though on the other hand this will also not have the benefits of the implicit optimization which #508 has, because going through several fallback fonts _can_ take time - especially for codepoints which many fonts don'e have glyphs for. We do occasionally get reports in mpv that OSD/subs rendering hangs the player shortly (dropped frames, etc), so avoiding these fallbacks for glyphs which are not ubiquitous and/or not needed anyway has an interesting potential here.

**Comment by khaledhosny** [comment]
Control characters that require no glyph in the font, should be kept in the same run as the surrounding text as they might influence shaping. Hard buzz knows about such code points and will behave correctly whether they are covered by the font or not.

A better don't fallback approach is to shape first with HarfBuzz then reshape runs of glyphs that were not supported by the font (this lets HarfBuzz handle these Default_Ignorable code points, as well as do NFC/NFD normalization based on what form is supported by the font). Chrome does a variation of this (with some optimizations).

**Comment by avih** [comment]
> A better don't fallback approach

This sounds nicer, yes, but I think this would need better knowledge of the libass dataflow than I have currently.

Do you want to take over making this happen? I don't mind closing #508. My only concern is that someone addresses it. I don't really carfe who.

**Comment by astiob** [comment]
That sounds interesting. Do you happen to have a link at hand that shows Chrome’s code?

**Comment by astiob** [comment]
Found a description in https://github.com/chromium/chromium/blob/master/third_party/blink/renderer/platform/fonts/README.md. Presumably, the relevant code is in the same place. I’ll report back when I learn something useful.

**Comment by avih** [comment]
> Do you want to take over making this happen?

Apologies, I accidentally thought this was written by @astiob (though of course you can take over if you feel like it ;) )

**Comment by astiob** [comment]
In short:
1. feed text to HarfBuzz claiming it all to be the same font;
2. re-shape any clusters (!) that contain .notdef glyphs;
3. repeat.

But this is not without issues. In particular, U+3000 needs special handling because HarfBuzz is eager to synthesize a glyph for it, which can ruin layout by using wrong metrics and disabling ligatures: https://github.com/harfbuzz/harfbuzz/issues/2889.

I’ve also looked into Gecko for comparison. It splits runs by font much as we do before passing them into HarfBuzz, but much like we’ve come to realize in #508, for some code points ([listed within Gecko itself](https://searchfox.org/mozilla-central/source/intl/unicharutil/util/nsUnicodeProperties.cpp#159)), it bypasses normal font splitting and adds them to the preceding shaping run. Curiously, these special code points include ZWJ and ZWNJ, _don’t_ include U+2060 or U+200B, but _do_ include e. g. Unicode marks and emoji skin colour selectors.

**Comment by avih** [comment]
> bypasses normal font splitting and adds them to the preceding shaping run. Curiously, these special code points include ZWJ and ZWNJ, _don’t_ include U+2060 or U+200B

My guess would be that the existing codepoints were added mainly ad-hoc, in response to bug report.

If indeed it looks for U+2060 at the font and breaks the run if the glyph is missing (we already know it's missing in Comic Sans and few more), and assuming there are cases where a broken run results in incorrect rendering, then it should be fairly trivial to synthesize a test case where it behaves incorrectly.

I'd even think that it would be easier to synthesize a bad use case with the RTL directional control, because as far as I can tell they're also excluded, and their impact when outside the run is probably bigger than with WJ.

**Comment by avih** [comment]
> >  is there a better way to escape backslash in libass other than artificially inserting some non-printable codepoint after backslash?
>
> Nope. We have an issue somewhere on the tracker for this, but there’s no easy or backwards-compatible way to add this.

> For reference, mpv internally uses a "trick" to communicate ass escaping between different parts at the code (e.g. when an mpv script command wants to display some text but only wants some of it to be ass-escaped): for an abitrary input string, all characters between the byte `0xfe` and `0xfd` (both invalid bytes in UTF-8) go through mpv ass escaping.

> Edit: also, I don't know how much U+2060 (or U+200B) is common in the wild. I think it's only common in mpv with libass because it's the best way to escape arbitrary text, but ultimately it's a hack, and not the optimal approach to escape arbitrarily.

Possibly slightly off topic for this issue, but not off-topic for the origin of the issue (the need to use hacks to escape):

I've now looked at Private-Use Characters - http://www.unicode.org/faq/private_use.html and https://en.wikipedia.org/wiki/Private_Use_Areas and I think libass could publicly declare and use such codepoints in its API to define escaping in a way similar to what mpv does internally.

**Comment by khaledhosny** [comment]
The issue with synthesized spaces in Chrome is probably because of the way Chrome does fallback. If I were to do this, I’d always keep any space with the text around it when falling back to a different font, because even if the first font had the space glyph it might not be suitable for the fallback font (especially if the fallback is for a different script).
